autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/importing/blob/WebViewBlobDownloader.kt

```diff
@@ -0,0 +1,163 @@
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+/**
+ * This interface provides the ability to add modern blob download support to a WebView.
+ */
+interface WebViewBlobDownloader {
+
+    /**
+     * Configures a web view to support blob downloads, including in iframes.
+     */
+    suspend fun addBlobDownloadSupport(webView: WebView)
+
+    /**
+     * Requests the WebView to convert a blob URL to a data URI.
+     */
+    suspend fun convertBlobToDataUri(blobUrl: String)
+
+    /**
+     * Stores a reply proxy for a given location.
+     */
+    suspend fun storeReplyProxy(
+        originUrl: String,
+        replyProxy: JavaScriptReplyProxy,
+        locationHref: String?,
+    )
+
+    /**
+     * Clears any stored JavaScript reply proxies.
+     */
+    fun clearReplyProxies()
+}
+
+@ContributesBinding(FragmentScope::class)
+class WebViewBlobDownloaderModernImpl @Inject constructor(
+    private val webViewCapabilityChecker: WebViewCapabilityChecker,
+) : WebViewBlobDownloader {
+
+    // Map<String, Map<String, JavaScriptReplyProxy>>() = Map<Origin, Map<location.href, JavaScriptReplyProxy>>()
+    private val fixedReplyProxyMap = mutableMapOf<String, Map<String, JavaScriptReplyProxy>>()
+
+    @SuppressLint("RequiresFeature")
+    override suspend fun addBlobDownloadSupport(webView: WebView) {
+        if (isBlobDownloadWebViewFeatureEnabled()) {
+            WebViewCompat.addDocumentStartJavaScript(webView, script, setOf("*"))
+        }
+    }
+
+    @SuppressLint("RequiresFeature")
+    override suspend fun convertBlobToDataUri(blobUrl: String) {
+        for ((key, proxies) in fixedReplyProxyMap) {
+            if (sameOrigin(blobUrl.removePrefix("blob:"), key)) {
+                for (replyProxy in proxies.values) {
+                    replyProxy.postMessage(blobUrl)
+                }
+                return
+            }
+        }
+    }
+
+    override suspend fun storeReplyProxy(
+        originUrl: String,
+        replyProxy: JavaScriptReplyProxy,
+        locationHref: String?,
+    ) {
+        val frameProxies = fixedReplyProxyMap[originUrl]?.toMutableMap() ?: mutableMapOf()
+        // if location.href is not passed, we fall back to origin
+        val safeLocationHref = locationHref ?: originUrl
+        frameProxies[safeLocationHref] = replyProxy
+        fixedReplyProxyMap[originUrl] = frameProxies
+    }
+
+    private fun sameOrigin(
+        firstUrl: String,
+        secondUrl: String,
+    ): Boolean {
+        return kotlin.runCatching {
+            val firstUri = Uri.parse(firstUrl)
+            val secondUri = Uri.parse(secondUrl)
+
+            firstUri.host == secondUri.host && firstUri.scheme == secondUri.scheme && firstUri.port == secondUri.port
+        }.getOrNull() ?: return false
+    }
+
+    override fun clearReplyProxies() {
+        fixedReplyProxyMap.clear()
+    }
+
+    private suspend fun isBlobDownloadWebViewFeatureEnabled(): Boolean {
```

Comment:
I'd like to keep this, just because that's also what it's called in `BrowserTabFragment` and the less this code differs from the original the easier it'll be to remove the duplicate code one day (assuming we get to doing that).

Let me know if you feel strongly it needs renamed.